DESIGN.md

```diff
@@ -0,0 +1,237 @@
+# ALPS
+Because every Go package requires a clever 4-letter acronym.
+
+## Structure
+Minimum Go version will be 1.16 because that's what's installed on my machine, and Go modules will be used for dependency 
+management. The overall architecture of the example program implementing the ALPS library will be 3 separate 
+binaries. It could be done in one binary, with the job worker also providing the HTTPS server, and providing the
+CLI commands by checking to see if the job worker is already running first before launching a new worker. However,
+I think that designing it with this structure provides for a clean, simple implementation, provides a separation 
+of concerns, and makes future extensibility easier. For example, it's likely that a feature such as scheduling 
+jobs which occur in the future or are recurring would be developed. In that case, it would be best to have the 
+job worker able to operate entirely independently of the HTTP service to isolate it from failures.
+
+- *pkg*
+    - *ipc*: types and methods required for inter-process communication.
+    - *log*: implement a common logging interface accross all components. This will not be developed from scratch,
+    just adapt the logger used in other projects
+    - *alps*: library code for creating and listening to Arbitrary Linux Process Service.
+- *cmd*
+    - *api*: HTTP API server using mTLS authentication with ACL authorization. Will communicate with the job 
+    worker via RPC. Authentication and authorization will be done via middleware. Any endpoint will return 
+    `{"status": 401, "error": "you are not authorized to perform that action"}` if the request does not have a recognized
+    user.
+        - **POST /start** starts a process and returns an ID to use for getting output and control
+            - Request: 
+            ```json
+            {"job": "ls -lh /var/log"}
+            ```
+            - Success response: 
+            ```json
+            {
+                "status": 200, 
+                "jobID": 123, 
```

Comment:
Since there is no persistence to worry about collisions, and I am mostly avoiding use of 3rd party libraries, it will just be a counter tracked by the job worker. In a production environment, this should be something like a UUID.

Reply:
Yes, in the updated trade-offs section I mention how a sequential ID could be vulnerable and that it should be a proper UUID. I think that is mitigated by the fact that the job has an owner and so simply knowing the ID of the job doesn't necessarily grant access, but there's no reason not to use a UUID besides avoiding 3rd-party libraries. (Or perhaps if this was an insanely high volume application and there was a need to cut overhead anywhere possible)

I could use some other scheme like just generating a random string or number, but then I would have to check for collisions even on the objects in memory. That would dramatically increase overhead when creating a job however, especially if there was a persistence layer.

Reply:
I think UUIDv4 is fine, it's just a randomly generated large number anyway. Then you don't have to worry about leaking any information.

Reply:
Ok, will do.